### PR TITLE
[Core] Adding delaunator utilities-Using triangle

### PR DESCRIPTION
**Pull request by loumalouomega**
In the future I will add an utility for easy access (or maybe more Kratos focussed, it is already very easy to use),  and I will add compatibility with the mortar segmentation utility.

To use, just build a std::vector<double> with the coordinates (2D) of the points.

### CODE REVIEW

**Comment by philbucher** [comment]
looks ok to me but adding new external libraries requires approval from @KratosMultiphysics/technical-committee I think
ping @KratosMultiphysics/technical-committee

**Comment by philbucher** [comment]
Take care, you are changing unrelated things!

**Comment by loumalouomega** [comment]
> Take care, you are changing unrelated things!

It is our repository, @tteschemacher discovered that the library could be included only once

**Comment by loumalouomega** [comment]
> > Take care, you are changing unrelated things!
> 
> It is our repository, @tteschemacher discovered that the library could be included only once

I will open an issue on the original repo

**Comment by philbucher** [comment]
I meant the changes in the Apps!

**Comment by loumalouomega** [comment]
> I meant the changes in the Apps!

This is probably a merge error

**Comment by loumalouomega** [comment]
Ok, i fixed the conflicts, sorry

**Comment by loumalouomega** [comment]
@tteschemacher I added "utilities/delaunator_utilities.h" as an example of use

**Comment by oberbichler** [comment]
I am always deeply impressed by the speed of JavaScript...

**Comment by loumalouomega** [comment]
> I am always deeply impressed by the speed of JavaScript...

What do you mean?

**Comment by oberbichler** [comment]
> Library is ~10% faster then JS version for 1M uniform points

For a scripting language JavaScript does very well. I made similar experiences with the NURBS.

**Comment by loumalouomega** [comment]
> > Library is ~10% faster then JS version for 1M uniform points
> 
> For a scripting language JavaScript does very well. I made similar experiences with the NURBS.
Yes, indeed

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee has three questions: @loumalouomega 

1. What would be the added feature of this respect to the triangle library?
2. Why it should go to the core and not to i.e meshing application?
3. What is the compiler requirement of the library?

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/technical-committee has three questions: @loumalouomega
> 
>     1. What would be the added feature of this respect to the triangle library?
> 
The library is in theory faster (I need to prepare the benchamrk on Kratos, but there is one on the original repository). The library is easier to use (and is real C++, not pure C). As is C++ the memory management is less error prone than C.
>     2. Why it should go to the core and not to i.e meshing application?
> 
This library (or any library equivakent) was required by @tteschemacher for the Isogeometric applications. Additionally I want to incorporate it into the mortar integration (in Core)
>     3. What is the compiler requirement of the library?
In theory is doesn't even require modern C++, so basically any C++ compiler

File: kratos/python/add_utilities_to_python.cpp
Comment (philbucher) [inline kratos/python/add_utilities_to_python.cpp]: I would say this should be exposed as a submodule, like [here](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_utilities_to_python.cpp#L914-L989)

**Comment by RiccardoRossi** [comment]
I have a number of questions:

1 - can this do CDT (Constrained Delaunay Triangulation) like Triangle
2 - can u assign colors to different parts of the domain and retain them?
3 - what options are available for this lib?
4 - is it really faster? 
5 - is it robust? (particularly for CDT provided it does it)

I am open to the possibility of moving to this, nevertheless we should then remove triangle. what i am against is having in the repository two libs doing the same thing.

**Comment by loumalouomega** [comment]
> I have a number of questions:
> 
> 1 - can this do CDT (Constrained Delaunay Triangulation) like Triangle

It is just standard Delaunay Triangulation

> 2 - can u assign colors to different parts of the domain and retain them?

The library is super super basic, you just give point, and returns triangles. Something like that must be implemented in our side. But the point of this library is for a much more basic approach

> 3 - what options are available for this lib?

Just one, give points, return triangles

> 4 - is it really faster?

Yes, but I did not have time to do properly the benchmarks. Besides triangle some time give memory errors

> 5 - is it robust? (particularly for CDT provided it does it)

It does what it does. It is a straightforward library, with only one purpose, it does fast, it consumes almost nothing (constructing it, everything except the triangles generated are references to the given values, everything happens in the constructor)

> I am open to the possibility of moving to this, nevertheless we should then remove triangle. what i am against is having in the repository two libs doing the same thing.

The idea of this library is not to replace triangle, but to be used there were a basic Delaunay is enough, and using triangle is to much consuming and implies some difficulties in order to be used (I have experiences with triangle giving me memory errors)

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee observes that the library is not maintain for last 10 month. So It would be better to keep track of it for a while to see if is reliable before adding it to the core.

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/technical-committee observes that the library is not maintain for last 10 month. So It would be better to keep track of it for a while to see if is reliable before adding it to the core.

I understand the argument, but I want to point that triangle has not being updated since 2005: 
https://www.cs.cmu.edu/~quake/triangle.html

**Comment by roigcarlo** [comment]
For the moment @KratosMultiphysics/technical-committee preffers to keep using the triangle library. We will repopen the discussion in the future in case the library improves.

This means that it will not be added to the core, but you can use it in other applications.

**Comment by loumalouomega** [comment]
The triangle library is broken in windows according to @AlejandroCornejo  and @ipouplana. Additionally as I said the objective of this mesh is simpler than the triangle. I don't need a whole CDT, I need a simple and fast delaunator to construct a set of triangles from a points cloud

**Comment by loumalouomega** [comment]
I updated the delaunator utilities so by default consider triangle instead of the new library, which is now an external dependency. To use:

~~~
-DINCLUDE_DELAUNATORCPP=ON                                                                        \
-DDELAUNATORCPP_INCLUDE_DIR="/home/vicente/src/delaunator-cpp/include/"                           \
~~~

I will check the windows compilation that looks like it fails. But first I will measure the execution times in order to compare velocities.

**Comment by philbucher** [comment]
> I updated the delaunator utilities so by default consider triangle instead of the new library, which is now an external dependency. To use:
> 
> ```
> -DINCLUDE_DELAUNATORCPP=ON                                                                        \
> -DDELAUNATORCPP_INCLUDE_DIR="/home/vicente/src/delaunator-cpp/include/"                           \
> ```

If this will be the way to go then I suggest to use the same mechanism as for AMatrix, then you only need one CMake argument:
If `DDELAUNATORCPP_INCLUDE_DIR` is defined, then use the delaunator, otherwise not

File: kratos/CMakeLists.txt
Comment (philbucher) [inline kratos/CMakeLists.txt:200]: this is not related/conflicting with #5538?

File: kratos/python/add_utilities_to_python.cpp
Comment (philbucher) [reply kratos/python/add_utilities_to_python.cpp]: ping

**Comment by loumalouomega** [comment]
> If this will be the way to go then I suggest to use the same mechanism as for AMatrix, then you only need one CMake argument:
> If `DDELAUNATORCPP_INCLUDE_DIR` is defined, then use the delaunator, otherwise not

I prefer this method, otherwise I need to change something in the only line

File: kratos/CMakeLists.txt
Comment (loumalouomega) [reply kratos/CMakeLists.txt:200]: Nope, the files from #5538 are not even imported, linked. Can be removed safely

**Comment by loumalouomega** [comment]
OK, after properly check the computation time, the reduction of this library respect triangle is less than a 5%, so no significant

**Comment by loumalouomega** [comment]
> OK, after properly check the computation time, the reduction of this library respect triangle is less than a 5%, so no significant

In fact for the small examples (the ones that concern me), triangle is slightly faster

**Comment by loumalouomega** [comment]
@tteschemacher Use the functions from delaunator_utilities. It works in the same way the delaunator does (even when triangle is used). You provide a vector of doubles and it returns the connectivity vector, this way you can reuse whatever you already implemented with many modification, sorry for the inconveniences

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee does not see any advantage in the delaunator anymore, (Since it is not even faster) and do not agree with adding it to the core even optionally. 

We recommend to reuse the delaunator utilities to improve the API for triangle.

Meanwhile, what is the problem with triangle in Windows? In case that we cannot make it work in windows we may reopen this discussion.

**Comment by loumalouomega** [comment]
> @KratosMultiphysics/technical-committee does not see any advantage in the delaunator anymore, (Since it is not even faster) and do not agree with adding it to the core even optionally.
> 
Delaunator is not added, not a single file of Delaunator is included in Kratos.  I just added the option to use it instead of triangle because triangles does not work in Windows anymore @AlejandroCornejo 

> We recommend to reuse the delaunator utilities to improve the API for triangle.
> 

This is what is done right now, it uses triangle by default and in case of including delaunator it is used instead

> Meanwhile, what is the problem with triangle in Windows? In case that we cannot make it work in windows we may reopen this discussion.

It crashes, @AlejandroCornejo please open an issue

**Comment by RiccardoRossi** [comment]
as we understand the triangle library is now working correctly in windows. Can we close this?

**Comment by loumalouomega** [comment]
> 
> 
> as we understand the triangle library is now working correctly in windows. Can we close this?

If you take a look to the changes included in this PR, (and the new title of the PR), this does not add anymore the new library. It adds utilities that simplify the usage of triangle (and if you want you can use delaunator, but must be in your computer, not included in Kratos). this is because triangle is not very intuitive to use, as far as you need to do a minimal of memory management. At the end the important thing are the utilities added. please check the changes introduced in the PR

**Comment by loumalouomega** [comment]
Ping

File: kratos/utilities/delaunator_utilities.h
Comment (philbucher) [inline kratos/utilities/delaunator_utilities.h]: @loumalouomega here is still delaunator

File: kratos/utilities/delaunator_utilities.cpp
Comment (philbucher) [inline kratos/utilities/delaunator_utilities.cpp]: here it is still

File: kratos/CMakeLists.txt
Comment (philbucher) [inline kratos/CMakeLists.txt]: here it is still

File: kratos/utilities/delaunator_utilities.h
Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.h]: But it is not mandatory, only if included, do you want me to absolutely remove it?

File: kratos/utilities/delaunator_utilities.cpp
Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.cpp]: It is not possible to define it as an external dependency, not mandatory?

File: kratos/CMakeLists.txt
Comment (loumalouomega) [reply kratos/CMakeLists.txt]: idem

**Comment by loumalouomega** [comment]
@philbucher I removed all reference to delaunator

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee changes requested applied

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee changes requested applied

**Comment by loumalouomega** [comment]
@philbucher comments were adressed

File: kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp
Comment (philbucher) [inline kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp:37]: shouldn't you also check the connectivities?

File: kratos/utilities/delaunator_utilities.cpp
Comment (RiccardoRossi) [inline kratos/utilities/delaunator_utilities.cpp:16]: this cannot be defined globally...as a minimum it must be undefined at the end of the file...

File: kratos/utilities/delaunator_utilities.cpp
Comment (RiccardoRossi) [inline kratos/utilities/delaunator_utilities.cpp]: i would include the include ward in the triangle.h file (maybe #include once)

**Review by RiccardoRossi — COMMENTED** [review]
@KratosMultiphysics/technical-committee must decide if allowing to link triangle lib in the core.

File: kratos/utilities/delaunator_utilities.cpp
Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.cpp]: OK

File: kratos/utilities/delaunator_utilities.cpp
Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.cpp:16]: OK

File: kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp
Comment (loumalouomega) [reply kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp:37]: I did that when the utilities were using triangle and the delaunator library, but after removing delaunator I just check the total area (see the checks), that is the relevant check, because the connectivities may change depending of the Delaunay parameters considered, but at the end the total area is the correct thing to check (well, is the alpha is to small no triangles are going to be detected, but we will not consider this case)

File: kratos/utilities/delaunator_utilities.cpp
Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.cpp]: I just checked and this is exactly the same as using #include once, in fact #include once is not supported by all compilers

File: kratos/utilities/delaunator_utilities.cpp
Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.cpp:16]: It cannot b e defined

Comment (RiccardoRossi) [reply kratos/utilities/delaunator_utilities.cpp:16]: ? why not?

Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.cpp:16]: In any case this is not defined globally, this is a cpp file (not a header), an it is the only way to make triangle to work. All implementations of triangle work that way.

File: kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp
Comment (philbucher) [reply kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp:37]: ok for me, but there is no check for Area, only number of elements

File: kratos/utilities/delaunator_utilities.cpp
Comment (philbucher) [inline kratos/utilities/delaunator_utilities.cpp]: this is not supported by all compilers, can you please do what we usually do

File: kratos/utilities/delaunator_utilities.cpp
Comment (philbucher) [reply kratos/utilities/delaunator_utilities.cpp:16]: you can do `#undef REAL` at the end of the file

File: kratos/utilities/delaunator_utilities.cpp
Comment (philbucher) [reply kratos/utilities/delaunator_utilities.cpp]: the point is that the guard does not have to be added every time triangle is included

File: kratos/utilities/delaunator_utilities.cpp
Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.cpp]: I did what we usually do before, i was confused with @RiccardoRossi comment

File: kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp
Comment (loumalouomega) [reply kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp:37]: OK, you mean in this test

File: kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp
Comment (loumalouomega) [inline kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp:102]: On this tests the area is check

File: kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp
Comment (loumalouomega) [reply kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp:37]: I can add an area to this test too

Comment (loumalouomega) [reply kratos/tests/cpp_tests/utilities/test_delaunator_utilities.cpp:37]: I added a test based on area. Funny thing is that I added directly from GitHub and it worked (I am evil ;))

File: kratos/utilities/delaunator_utilities.cpp
Comment (philbucher) [reply kratos/utilities/delaunator_utilities.cpp:16]: ok this is done

File: kratos/utilities/delaunator_utilities.cpp
Comment (philbucher) [reply kratos/utilities/delaunator_utilities.cpp]: so once you did this I think this is ready to go in

Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.cpp]: What I am supposed to do?, this is the gaurd we use in all files

File: kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp
Comment (philbucher) [reply kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp:102]: ok

File: kratos/utilities/delaunator_utilities.cpp
Comment (philbucher) [reply kratos/utilities/delaunator_utilities.cpp]: yes, the point is that we DO NOT WANT this guard in ALL THE FILES but instead only once in [triangle.h](https://github.com/KratosMultiphysics/Kratos/blob/master/external_libraries/triangle/triangle.h)

Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.cpp]: OK, I am not very pro modify external libraries, but as triangle has been the same since forever I think we can live with that

Comment (philbucher) [reply kratos/utilities/delaunator_utilities.cpp]: that was also our thinking

File: kratos/CMakeLists.txt
Comment (philbucher) [inline kratos/CMakeLists.txt]: should this be here?
I don't see this file in this PR

File: kratos/utilities/delaunator_utilities.h
Comment (philbucher) [inline kratos/utilities/delaunator_utilities.h]: just for the record (i.e. leave it as is)
this would IMO be an appropriate use for forward declaration, since then this heavy header could be moved to the cpp

File: kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp
Comment (philbucher) [inline kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp]: is this still needed?
also in other places

File: kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp
Comment (philbucher) [inline kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp:140]: Here you could also use std::array, might be slightly more efficient

**Review by philbucher — COMMENTED** [review]
very minor comments, otherwise ok for me now

@KratosMultiphysics/technical-committee feel free to approve if you also agree

File: kratos/utilities/delaunator_utilities.h
Comment (loumalouomega) [reply kratos/utilities/delaunator_utilities.h]: OK

File: kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp
Comment (loumalouomega) [reply kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp]: I can clean it

File: kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp
Comment (loumalouomega) [reply kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp:140]: Okay

File: kratos/CMakeLists.txt
Comment (loumalouomega) [reply kratos/CMakeLists.txt]: Duplicated

File: kratos/utilities/delaunator_utilities.h
Comment (philbucher) [inline kratos/utilities/delaunator_utilities.h]:
```suggestion
// forward declaring ModelPart to be avoid including heavy header here
```

File: kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp
Comment (loumalouomega) [reply kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp:140]: The problem with std::array is that implies to define it as template, and I don't like that...

Comment (philbucher) [reply kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp:140]: you mean the size?
I thought there was a way that the std::array determines this from the intializer list
anyway, not important

Comment (loumalouomega) [reply kratos/tests/cpp_tests/external_libraries/test_delaunator.cpp:140]: https://en.cppreference.com/w/cpp/container/array This requires not only the type, but the size which is fixed, so it can be done, but templating the method and moving all triangle stuff to the header

**Review by RiccardoRossi — DISMISSED** [review]
ok, approving on behalf of the @KratosMultiphysics/technical-committee

**Review by roigcarlo — DISMISSED** [review]
Prevent

**Comment by RiccardoRossi** [comment]
ouch, we just realized that the MeshingApplication should be changed since it is linking the triangle too. (change is minimal but it is needed for the MeshingApp to keep working)

**Comment by loumalouomega** [comment]
> 
> 
> ouch, we just realized that the MeshingApplication should be changed since it is linking the triangle too. (change is minimal but it is needed for the MeshingApp to keep working)

I already updated, see the files changed (also Delaunator App)

**Comment by RiccardoRossi** [comment]
you did not change the CMakeLists of the app

**Comment by loumalouomega** [comment]
> 
> 
> you did not change the CMakeLists of the app

Ah, Ok, you are right

**Comment by loumalouomega** [comment]
@RiccardoRossi I removed triangle linking in MeshingApplication

**Comment by loumalouomega** [comment]
@roigcarlo I attended the suggestions

**Comment by roigcarlo** [comment]
:+1: No block anymore, but I would wait for the rest of the @KratosMultiphysics/technical-committee to approve.

**Review by roigcarlo — APPROVED** [review]
:+1:

**Comment by loumalouomega** [comment]
Thanks

**Pull request merged**
